core/src/main/java/jenkins/model/PeepholePermalink.java

```diff
@@ -94,115 +90,216 @@ protected File getPermalinkFile(Job<?, ?> job) {
+     * There is one line per cached permalink, in the format {@code lastStableBuild 123}
+     * or (for a negative cache) {@code lastFailedBuild -1}.
      */
-    private Run<?, ?> find(Run<?, ?> b) {
-        //noinspection StatementWithEmptyBody
-        for ( ; b != null && !apply(b); b = b.getPreviousBuild())
-            ;
-        return b;
-    }
+    @Restricted(NoExternalUse.class)
+    @Extension(ordinal = -1000)
+    public static final class DefaultCache implements Cache {
+
+        /**
+         * JENKINS-22822: avoids rereading caches.
+         * Top map keys are {@code builds} directories.
+         * Inner maps are from permalink name to target.
+         * Synchronization is first on the outer map, then on the inner.
+         */
+        private final Map<File, Map<String, Known>> caches = new HashMap<>();
 
-    private static @NonNull Map<String, Integer> cacheFor(@NonNull File buildDir) {
-        synchronized (caches) {
-            Map<String, Integer> cache = caches.get(buildDir);
-            if (cache == null) {
-                cache = load(buildDir);
-                caches.put(buildDir, cache);
+        @Override
+        public PermalinkTarget get(Job<?, ?> job, String id) {
+            var cache = cacheFor(job.getBuildDir());
+            synchronized (cache) {
+                var cached = cache.get(id);
+                return cached != null ? cached : UNKNOWN;
             }
-            return cache;
         }
-    }
 
-    private static @NonNull Map<String, Integer> load(@NonNull File buildDir) {
-        Map<String, Integer> cache = new TreeMap<>();
-        File storage = storageFor(buildDir);
-        if (storage.isFile()) {
-            try (Stream<String> lines = Files.lines(storage.toPath(), StandardCharsets.UTF_8)) {
-                lines.forEach(line -> {
-                    int idx = line.indexOf(' ');
-                    if (idx == -1) {
-                        return;
-                    }
+        @Override
+        public void put(Job<?, ?> job, String id, Known target) {
+            File buildDir = job.getBuildDir();
+            var cache = cacheFor(buildDir);
+            synchronized (cache) {
+                cache.put(id, target);
+                File storage = storageFor(buildDir);
+                LOGGER.fine(() -> "saving to " + storage + ": " + cache);
+                try (AtomicFileWriter cw = new AtomicFileWriter(storage)) {
                     try {
-                        cache.put(line.substring(0, idx), Integer.parseInt(line.substring(idx + 1)));
-                    } catch (NumberFormatException x) {
-                        LOGGER.log(Level.WARNING, "failed to read " + storage, x);
+                        for (var entry : cache.entrySet()) {
+                            cw.write(entry.getKey());
+                            cw.write(' ');
+                            cw.write(Integer.toString(entry.getValue() instanceof Cache.Some some ? some.number : -1));
```

Comment:
Note that this can be written more clearly in Java 21:

```diff
--- core/src/main/java/jenkins/model/PeepholePermalink.java
+++ core/src/main/java/jenkins/model/PeepholePermalink.java
@@ -248,7 +248,10 @@ public abstract class PeepholePermalink extends Permalink implements Predicate<R
                         for (var entry : cache.entrySet()) {
                             cw.write(entry.getKey());
                             cw.write(' ');
-                            cw.write(Integer.toString(entry.getValue() instanceof Cache.Some some ? some.number : -1));
+                            cw.write(Integer.toString(switch (entry.getValue()) {
+                                case Some some -> some.number;
+                                case None none -> -1;
+                            }));
                             cw.write('\n');
                         }
                         cw.commit();
```